libs/execution_context/ebpf_native.c

```diff
@@ -0,0 +1,1219 @@
+    }
+
+    return result;
+}
+
+ebpf_result_t
+ebpf_native_get_count_of_programs(_In_ const GUID* module_id, _Out_ size_t* count_of_programs)
+{
+    ebpf_result_t result = EBPF_SUCCESS;
+    ebpf_lock_state_t state = 0;
+    ebpf_native_t** native_module = NULL;
+
+    // Find the native entry in hash table.
+    state = ebpf_lock_lock(&_ebpf_native_client_table_lock);
+    result = ebpf_hash_table_find(_ebpf_native_client_table, (const uint8_t*)module_id, (uint8_t**)&native_module);
+    if (result != EBPF_SUCCESS) {
+        result = EBPF_OBJECT_NOT_FOUND;
+        goto Done;
+    }
+
+    *count_of_programs = _ebpf_native_get_count_of_programs(*native_module);
+
+Done:
+    ebpf_lock_unlock(&_ebpf_native_client_table_lock, state);
+    return result;
+}
+
+ebpf_result_t
+ebpf_native_get_count_of_maps(_In_ const GUID* module_id, _Out_ size_t* count_of_maps)
+{
+    ebpf_result_t result = EBPF_SUCCESS;
+    ebpf_lock_state_t state = 0;
+    ebpf_native_t** native_module = NULL;
+
+    // Find the native entry in hash table.
+    state = ebpf_lock_lock(&_ebpf_native_client_table_lock);
+    result = ebpf_hash_table_find(_ebpf_native_client_table, (const uint8_t*)module_id, (uint8_t**)&native_module);
+    if (result != EBPF_SUCCESS) {
+        result = EBPF_OBJECT_NOT_FOUND;
+        goto Done;
+    }
+
+    *count_of_maps = _ebpf_native_get_count_of_maps(*native_module);
+
+Done:
+    ebpf_lock_unlock(&_ebpf_native_client_table_lock, state);
+    return result;
+}
+
+ebpf_result_t
+ebpf_native_unload(_In_ const GUID* module_id)
+{
+    ebpf_result_t result;
+    ebpf_lock_state_t state = 0;
+    ebpf_lock_state_t native_state = 0;
+    bool lock_acquired = false;
+    bool module_lock_acquired = false;
+    ebpf_native_t** existing_native_module = NULL;
+    ebpf_native_t* native_module = NULL;
+    wchar_t* service_name = NULL;
+    size_t service_name_length;
+    bool unload_module = false;
+
+    // Find the native entry in hash table.
+    state = ebpf_lock_lock(&_ebpf_native_client_table_lock);
+    lock_acquired = true;
+    result =
+        ebpf_hash_table_find(_ebpf_native_client_table, (const uint8_t*)module_id, (uint8_t**)&existing_native_module);
+    if (result != EBPF_SUCCESS) {
+        result = EBPF_OBJECT_NOT_FOUND;
+        goto Done;
+    }
+    native_module = *existing_native_module;
+    native_state = ebpf_lock_lock(&native_module->lock);
+    module_lock_acquired = true;
+    if (native_module->unloading) {
+        // If module is already unloading, skip unloading it again.
+        result = EBPF_SUCCESS;
+        goto Done;
+    }
+    native_module->unloading = true;
+    unload_module = true;
+
+    // It is possible that the module is also detaching at the same time and
+    // the module memory can be freed immediately after the hash table lock is
+    // released. Create a copy of the service name to use later to unload driver.
+    service_name_length = (wcslen(native_module->service_name) * 2) + 2;
```

Comment:
```suggestion
    service_name_length = (wcslen(native_module->service_name) + 1) * sizeof(wchar_t);
```
nit: Easier to understand why the 2